docs: design for multisite improvements (#10323)
rook/rook: design/ceph/object/ceph-rgw-support-multisite-for-external-cluster.md

---
# Support Realm/Zone/Zonegroup for external RGW server
target-version: release-X.X
---

# Feature Name
 Support Realm/Zone/Zonegroup for external RGW server

## Summary

> Rook supports handling for external RGW servers by assuming it is part of `default` zone/zonegroup. To handle consistently it is better handle this info, otherwise the radosgw apis may fail for such RGW servers.

Reviewer: Where does rook assume the default zone/zonegroup?

Author: Please refer [here](https://github.com/rook/rook/blob/master/pkg/operator/ceph/object/objectstore.go#L198)

### Goals

> If external RGW server is configured with zone/zonegroup/realm, allow equivalent CR creation in Rook.

Reviewer: For an external RGW, rook doesn't create any RGW pod. So why do we need to create these CRs in Rook for the external RGW case?

Author: My alternate approach to defining these directly is on the `ObjectStoreSpec`. I am still guessing the `realm` k8s secret still needs to be created since considering the situation in which one cluster is external and the secondary one is internal

> ---

Reviewer: No need for the separators in the header, those are just for the generated docs, not the design docs
```suggestion
```

# Improvements to RGW Multisite configuration
target-version: release-1.10

> ---

Reviewer: ```suggestion
```

# Feature Name

>  Improvements to RGW Multisite configuration

Reviewer: Could you summarize the gaps in the current multisite config? I'm having difficulty understanding why we need the features suggested in this design.

## Summary
Following are the major changes planned with this feature:
* For external RGW servers, have place holder for zone/zonegroup/realm settings in `GatewaySpec`.
* Provide option to add Endpoints to `zone` configuration.
* Encode the `Endpoint` information into `realm-system-user` secret using the RGW Kubernetes service name.

### Goals

> Improve usability of RGW mulitisite for customers and make it more friendly for disaster recovery scenarios. And planning to cover following use-cases:

Reviewer 2: ```suggestion
Improve usability of RGW multisite for customers and make it more friendly for disaster recovery scenarios. And planning to cover the following use-cases:
```

* two ceph object store clusters running in two different kubernetes cluster.

> * one ceph object store resides in kubernetes and another is external.

Reviewer: Whether or not the object stores are internal or external, why does it matter? Seems like RGW multisite should behave the same in either case. Although I'm not sure it makes sense for the admin of a rook cluster to configure RGW multisite on an external cluster that is managed by cephadm. The cluster administrator should be the one configuring multisite configs.

Reviewer 3: Agree with @travisn on the first comment. My understanding of external mode is that it's only meant to expose RGWs to things running inside of K8s. When Rook sets this up there is no ceph cluster (Mons and OSDs) it is setting up. Unless my understanding of external mode is inaccurate, whether or not the External RGWs have multisite configured and are syncing data between another cluster doesn't change the way the resource is consumed or viewed by an application.

Reviewer 3: @thotz Here are my comments on your following scenarios:
1. Does OBC and CephObjectStoreUser fail when the Multisite CRDs are created? If so do you know why? That is an issue worth fixing, but I don't see it as a multisite improvement just maybe some added config to those CRDs
2. In this scenario all you would have to do is add an endpoint of an RGW from the master zone in the external Ceph cluster to the `endpoint` in the `pull` section of the `CephObjectRealm` config, and then create the realm and the zone group CRs with the same name as the external Ceph cluster. That's it, as long as the configured endpoint can be reached from the k8s environment.
3. This scenario is practically the same as the 2nd one. The only thing that matters between syncing two clusters, regardless of whether the cluster is configured in K8s or not is that the RGWs are in the realm reachable from one another, and matching the names.

Author: > @thotz Here are my comments on your following scenarios:
> 
>     1. Does OBC and CephObjectStoreUser fail when the Multisite CRDs are created? If so do you know why? That is an issue worth fixing, but I don't see it as a multisite improvement just maybe some added config to those CRDs
> 

The simple or common use cases having multiple zones with belongs to a zonegroup work fine for obc/cephobjectstoreuser if the ceph cluster is internal.  
for RGW belongs to external ceph cluster this was falling see upstream issue https://github.com/rook/rook/issues/6342 and ODF bug https://bugzilla.redhat.com/show_bug.cgi?id=2050224. For the ODF he mentioned to add the `internal-rgw-service-name` to zone list and it worked for him. I am not quite sure about that

For situations with zone-groups(use cases with two distant ceph-cluster/k8s cluster), this will be a bit tricky, especially for OBC

>     2. In this scenario all you would have to do is add an endpoint of an RGW from the master zone in the external Ceph cluster to the `endpoint` in the `pull` section of the `CephObjectRealm` config, and then create the realm and the zone group CRs with the same names from the external Ceph cluster. That's it, as long as the configured endpoint can be reached from the k8s environment.
> 
>     3. This scenario is practically the same as the 2nd one. The only thing that matters between syncing two clusters, regardless of whether the cluster is configured in K8s or not is that the RGWs are in the realm reachable from one another, and match the names.

Yeah, I completely agree that the above use-case works fine, but my idea is to improve this, not reimplement the wheel like adding more details to realm secrets

Reviewer 4: @leseb you posted #6342. Do you remember more context about that that could assist us with this design? My understanding of that issue was that we should be able to create realm/zonegroup/zone resources to assist that. Or did you intend some other strategy?

Reviewer 2: AFAIR, https://github.com/rook/rook/issues/6342's goal was to adapt `NewMultisiteContext()` so that `getMultisiteForObjectStore()` reports the correct realm/zonegroup/zone even if the cluster is external. This code:

https://github.com/rook/rook/blob/1acc0621af652b37b10c224064ba8c80df5e02fe/pkg/operator/ceph/object/objectstore.go#L198-L204

needs to change.

> * one ceph object store resides in kubernetes and another is external.

Reviewer 4: Aren't these two cases essentially equivalent? In both cases, won't there be one CephObjectStore in external mode?

* External ceph object store configured with custom zone/zonegroup/realms.

## Proposal details
Currently for external RGW server, Rook sets realm/zone/zonegroup in ObjContext to ""/default/default respectively. So if the RGW server belongs to custom realm/zone/zonegroup, all the `radosgw-admin` commands related may fail. So for external RGW server introducing fields to define realm/zone/zonegroup and will be part of `GatewaySpec` in `CephObjectStoreSpec`. And Rook operator will also create Kubernetes secret for storing system user details.
```

> type ExternalMultisiteSpec struct {

Reviewer: We have CRDs for the multisite configuration: CephObjectRealm, CephObjectZoneGroup, and CephObjectZone. Seems like we should use those instead of redefined them in the gateway.

Author: Didn't used CRDs since it does not generate any resources, we are just referring to existing details about RGW configuration. Rook operator is filling up Objcontext with zone/zonegroup/realm details irrespective of external or internal RGW servers. That information needs to populate correctly and it passed `radosgw-admin` commands later on.

Reviewer 2: Same comment as Travis, I would expect each multisite CRDs to implement an external capability or something, so that `ObjContext` is filled correctly?

Author: @travisn @leseb how about adding `external` as `bool` in the CRDs of realm/zonegroup/zone?

Reviewer: The realm has a [pull.endpoint](https://github.com/rook/rook/blob/master/deploy/examples/object-multisite-pull-realm.yaml#L23-L29) that need to be set if we are pulling from an external rgw. Does this already represent the external cluster?

Author: @travisn : Yaa it will work. IMO I guess the major confusion is mixing up two different features/issues together.

First issue, if RGW server part of external ceph cluster is exposed as k8s service(I mean `cephobjectstore` creation for external RGW) which has custom zone/zonegroup settings(let's leave out multisite from the picture at the moment) then all operations related such as creation/deletion/updation to `OBC`, `cephobjectstoreuser`, `cephobjectstore` needs to work properly. There are two issues related to that described in https://github.com/rook/rook/pull/10323#discussion_r887587521.

The second issue, If I need to configure up an internal add cephobjectstore to already existing external cephobjectstore. The user can manually create the secret with system user details, fetch the endpoint and give it to `realm.pull.endpoint`. What is in my mind was if the user creates(user can always choose not to) `cephobjecstore` for the rgw server in the external ceph cluster, then we can ease the manual step by encoding details into the current k8s realm secret, so that user can just refer the k8s secret for all the details

Reviewer 4: Still trying to understand this myself. What I think I understand is that there are two cases: (1) the multisite case, and (2) the external case.

1. In the multisite case, we assume to have our Rook CephObjecStore in non-external mode. This is a CephCluster running in Kubernetes. The CephObjectRealm `spec.pull.endpoint` refers to a different Ceph object store (it may be a different Rook cluster or may be standalone Ceph, we don't care), and all the information in the realm is synchronized between the two clusters. We effectively have a local copy of an external object store.

2. The external case (that is bugged as noted in #6342) is when we have an external mode CephObjectStore. In this case we use CephObjectStore `spec.gateway.externalRgwEndpoints` to tell Rook to send all object-related commands to a given endpoint. That RGW endpoint may have any zone/zonegroup/realm configuration, and Rook needs to use that info to be able to create buckets and users on the external RGW. We are not creating a local copy of the external store, but instead we are sending all communications to the external RGW which is the only source of truth.

@thotz @alimaredia does this breakdown sound right to you two?

For the external case bugged by #6342, I have further questions:
1. Is it possible for Rook to query Ceph or the RGW via CLI/API calls to determine what zone/zonegroup/realm an RGW is in when given its `spec.gateway.externalRgwEndpoints`?
2. My experience using Ceph's CLI has been that I don't always need all that information to do the API calls. Can we get by with only zone (or only zone and realm)(or only zone and zonegroup) information?

    Realm string `json:"realm,omitempty"`
    Zone string `json:"zone,omitempty"`
    ZoneGroup string `json:"zoneGroup,omitempty"`
    SystemUser string `json:"systemUser,omitempty"`
}

type GatewaySpec struct {
..

    ExternalMultisiteSettings ExternalMultisiteSpec `json:"externalMultisiteSettings,omitempty"`
}
``` 
The `ZoneGroup` endpoint is updated with RGW Service IP than the resolvable RGW service name in Kubernetes cluster. The realm pull may not work on the above mentioned use-cases. The new approach is either user can define endpointi(like openshift routes) the `ZoneSpec`  struct and then ZoneGroup Endpoint will filled with Master Zone Endpoint or it will be update with `Endpoint` defined as similar to Object Bucket Claims/CephObjectStoreUsers and for external RGW server it is already defined in `ObjectStoreSpec`.
```
type ZoneSpec struct {
..
     RGWEndpoint string `json:"rgwEndpoint,omitempty"`
}

```
Also planning to include Endpoint as part of Realm secrets, so user can directly perform get `Endpoint` directly than fetch it via `radosgw-admin` commands.

> ## Open Questions [optional]

Reviewer: If there are no open questions, you can delete this section
